uint32[] number_of_observations